### PR TITLE
Feature: 🚀 Add InboxRule implement 

### PR DESCRIPTION
**Pull request by a76yyyy**
Feat ecederstrand/exchangelib#233

Example：

```Python
from exchangelib import (
    IMPERSONATION,
    Account,
    Build,
    Configuration,
    Credentials,
    Version,
)
from exchangelib.properties import Actions, Conditions, Exceptions, Rule

SERVER_DOMAIN = 'mail.example.com'
USERNAME = r'example\username'
PASSWORD = 'password123!'

TO_USER = 'user@example.com'
ACCESS_TYPE = IMPERSONATION


def test_inbox_rule():
    # Create Credentials object
    credentials = Credentials(username=USERNAME, password=PASSWORD)

    # Create Configuration object, set server address and credentials
    config = Configuration(
        server=SERVER_DOMAIN,
        credentials=credentials,
        version=Version(build=Build(15, 1)),
        auth_type="basic"
    )

    # Create Account object, set email address, access type and configuration
    account = Account(
        primary_smtp_address=TO_USER,
        access_type=ACCESS_TYPE,  # Choose one from (IMPERSONATION, DELEGATE)
        config=config
    )

    print("Rules of account before creation:", account.rules, '\n')

    # Create rule conditions, sender contains 'sender_example'
    conditions = Conditions(
        contains_sender_strings=[
            "sender_example",
        ],
    )

    # Create rule actions, automatically delete emails
    actions = Actions(
        delete=True
    )

    # Create Rule instance
    rule = Rule(
        display_name='test_exchangelib_rule',
        priority=1,
        is_enabled=True,
        conditions=conditions,
        exceptions=Exceptions(),
        actions=actions,
    )

    # Create rule
    rule_id = account.create_rule(rule)
    print("Rule:", rule)
    print("Created rule with ID:", rule_id, '\n')

    if rule_id:
        # Get rule list
        print("Get Rules of account after creation:", account.rules, '\n')

        # Modify rule
        rule.rule_id = rule_id
        rule.display_name = 'test_exchangelib_rule(modified)'
        res = account.set_rule(rule)
        print("Modified rule with ID:", rule.rule_id)
        print("Response:", res.text if hasattr(res, 'text') else res)
        print("Rules of account after modification:", account.rules, '\n')

        # Delete rule
        res = account.delete_rule(rule=rule)
        print("Deleted rule with ID:", rule.rule_id)
        print("Response:", res.text if hasattr(res, 'text') else res)
        print("Rules of account after deletion:", account.rules)


if __name__ == "__main__":
    test_inbox_rule()
```

### CODE REVIEW

**Comment by a76yyyy** [comment]
image

I did not find a cycle import

**Comment by ecederstrand** [comment]
Thanks for the contribution!

We'll need to add a test cate to this MR. Your example code is a really good start.

Also, I usually update the generated docs in a different MR because they get published immediately. I want to postpone that until a release hits PyPI. However, a new section in https://github.com/ecederstrand/exchangelib/blob/master/docs/index.md describing the new feature would be good.

File: exchangelib/account.py [hunk 1]
```diff
@@ -747,6 +756,69 @@ def delegates(self):
...
+    def create_rule(self, rule: Rule):
+        """Create an Inbox rule in a user's mailbox in the Exchange store.
+
+        :param rule: The rule to create with display_name as the key at least.
+        :return: The rule ID of the created rule. If failed, raise an error.
+        """
+        create_rule_service = CreateInboxRule(account=self)
+        create_rule_service.get(rule=rule, remove_outlook_rule_blob=True)
+        # After creating the rule, query all rules,
+        # find the rule that was just created, and return its ID.
```
Comment (ecederstrand) [inline exchangelib/account.py:791]: Doesn't the `CreateInboxRule` service return an ID of the created rule? All other services do that.

File: exchangelib/account.py [hunk 2]
```diff
@@ -747,6 +756,69 @@ def delegates(self):
...
+
+    def set_rule(self, rule: Rule):
+        """Modify an Inbox rule in a user's mailbox in the Exchange store.
+
+        :param rule: The rule to set with rule_id as the key at least.
+        :return: None if success, else raise an error.
+        """
+        return SetInboxRule(account=self).get(rule=rule)
+
+    def delete_rule(self, rule_name: Optional[str] = None, rule_id: Optional[str] = None):
```
Comment (ecederstrand) [inline exchangelib/account.py:806]: I think it's better if this method just takes a `Rule` object instead. We can let the user get a hold of the rule ID via `Account.get_rules()` if they only have the name.

**Review by ecederstrand — COMMENTED** [review]
Overall, solid work. Thanks!

File: exchangelib/account.py (hunk 1 above)
Comment (a76yyyy) [reply exchangelib/account.py:791]: According to https://learn.microsoft.com/en-us/exchange/client-developer/web-service-reference/updateinboxrules-operation#code-1, CreateInBoxrule Services does not return ID.

**Comment by a76yyyy** [comment]
> Thanks for the contribution!
> 
> We'll need to add a test cate to this MR. Your example code is a really good start.
> 
> Also, I usually update the generated docs in a different MR because they get published immediately. I want to postpone that until a release hits PyPI. However, a new section in https://github.com/ecederstrand/exchangelib/blob/master/docs/index.md describing the new feature would be good.

I have updated `index.md`, but I did not continue to update the generated docs.If you think this PR is ready to merge, you can ask me to generate the latest docs (or do it by myself)

File: exchangelib/account.py (hunk 1 above)
Comment (ecederstrand) [reply exchangelib/account.py:767]: Ah, thanks for the pointer to the docs. I'm surprised - EWS is usually better designed than this.

**Comment by ecederstrand** [comment]
I added a test case, made the API on Account a bit more consistent, and reverted the automatic generation of docs.

**Comment by kingfer30** [comment]
Hello, I would like to ask if there is an interface for creating junk rules through the API? 

I saw that the rule created by this rule is only valid in the rule list and not in the junk list.
image

By reading the code, I found a method "bulk_mark_as_junk", but the official document states that blocked emails need to already exist, and my scenario is that certain the person or domain is not sending me emails, but I need to configure it so that I receive emails properly and not get marked as junk.
image

That like this:
image

Is it possible to implement? I'm a newbie, very sorry if this question is stupid

**Comment by a76yyyy** [comment]
> By reading the code, I found a method "bulk_mark_as_junk", but the official document states that blocked emails need to already exist, and my scenario is that certain the person or domain is not sending me emails, but I need to configure it so that I receive emails properly and not get marked as junk. 

Maybe you can configure an Inbox Rule: When you receive a mail with a specified sending address, automatically move it into junk mail

**Pull request merged**